cmd/loader/generation.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"strconv"
+	"strings"
+	"sync/atomic"
+)
+
+type metricData struct {
+	count           uint64 // atomic
+	nameFormat      string
+	nameCardinality uint
+	tagCardinality  []uint
+	valueLimit      uint
+}
+
+type metricGenerator struct {
+	rnd *rand.Rand
+
+	counters metricData
+	gauges   metricData
+	sets     metricData
+	timers   metricData
+}
+
+func (md *metricData) genName(sb *strings.Builder, r *rand.Rand) {
+	sb.WriteString(fmt.Sprintf(md.nameFormat, r.Intn(int(md.nameCardinality))))
+	sb.WriteByte(':')
+}
+
+func (md *metricData) genTags(sb *strings.Builder, r *rand.Rand) {
+	if len(md.tagCardinality) > 0 {
+		sb.WriteString("|#")
+		for idx, c := range md.tagCardinality {
+			if idx > 0 {
+				sb.WriteByte(',')
+			}
+			sb.WriteString(fmt.Sprintf("tag%d:%d", idx, r.Intn(int(c))))
+		}
+	}
+	sb.WriteByte('\n')
+}
+
+func (mg *metricGenerator) nextCounter(sb *strings.Builder) {
+	atomic.AddUint64(&mg.counters.count, ^uint64(0))
```

Comment:
Can I ask why we are using the max unit64 value?
Is that just going to overflow the value?

Reply:
This is how you do an atomic subtract from an unsigned - `^0` === max unsigned === `-1` signed.

https://golang.org/pkg/sync/atomic/#AddUint64